tools/validate/validate.cpp

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CRT_SECURE_NO_WARNINGS
+#include <algorithm>
+#include <array>
+#include <assert.h>
+#include <filesystem>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+using namespace std;
+
+class BinaryFile {
+public:
+    explicit BinaryFile(const filesystem::path& filepath) {
+        m_file = _wfopen(filepath.c_str(), L"rb");
+
+        if (!m_file) {
+            fwprintf(stderr, L"Validation failed: %s couldn't be opened.\n", filepath.c_str());
+        }
+    }
+
+    [[nodiscard]] bool read_next_block(vector<unsigned char>& buffer) {
+        constexpr size_t BlockSize = 65536;
+
+        buffer.resize(BlockSize);
+
+        const size_t bytes_read = fread(buffer.data(), 1, BlockSize, m_file);
+
+        buffer.resize(bytes_read);
+
+        return !buffer.empty();
+    }
+
+    ~BinaryFile() {
+        if (fclose(m_file) != 0) {

```

Comment:
the fact fclose is allowed to return errors is an abomination. No change requested.